client/tests/integration/smartcontracts/validator_with_custom_token/src/lib.rs

```diff
@@ -56,22 +52,33 @@ mod token {
     pub struct CanControlDomainLives;
 }
 
-struct Validator {
-    verdict: Result,
-    block_height: u64,
-    host: iroha_wasm::Host,
-}
+default_validator!(
```

Comment:
Can this be a derive macro instead? Or probably two. Something like:

```rust
#[derive(Validate, Visit)]
// Tells `Visit` macro to fallback to `visit_register_domain()` function from this module
// instead of `iroha_wasm_validator::default::visit_register_domain()`
#[visit(custom(visit_register_domain))]
pub struct Validator {
    // Here goes probably the same fields for every validator,
    // But I don't think it's a big problem and prefer transparency
}

pub fn visit_register_domain(&mut Validator, authority: &AccountId, isi: Register<Domain) {
    // ...
}
```

Reply:
@mversic , what do you think?

Reply:
> @mversic , what do you think?

IMO your suggestion looks more complicated. Why would you enforce derive macro?

Reply:
I was thinking about going that route as well, but the biggest problem I've seen is that the visit functions should be inside an `impl Visit` block, and we can only have one that's generated by the macro. The idea of an attribute specifying a fallback has somehow slipped me. 😢 I'd say that with such a scheme the code looks a bit more cleaner and compact, and re: complicatedness I think it's only marginally worse, and only in terms of implementation, whereas user-facing experience is indeed more straightforward.

Reply:
@ilchu you will find that you can't alter struct's body with derive macro. Which means that you cannot hide those repeating fields like `verdict`. Also, how do you measure complicatedness? I find it quite incomprehensible to read the code written by @Arjentix whereas I had no problem in reading your original code

I'd like to hear @Erigara and @DCNick3 opinions here. Do you prefer function like macro or derive macro in this case?

Reply:
I like the `derive(Validate)`. To "hide" extra fields you can put them all into a struct `ValidatorBase` (name up for bike shed) defined in `iroha_validator`.

As for the `derive(Visit)`, I am not even sure why this is required, see https://github.com/hyperledger/iroha/pull/3896#issuecomment-1724995062.

Reply:
> Item-level function-like macros should be avoided IMO, they are often too much magic

I agree. `derive`-macros has explicit abilities and clearly stays to which trait they're attached to.

> Which means that you cannot hide those repeating fields like verdict

Yes, but `derive`-macro can panic if some field is missing

Reply:
I feel it's more of a question about explicitness and what we would want to provide to the end user. The current func-like macro solution hides more things, but allows arbitrary magic (and thus has more assumptions baked in too), making it more frustrating when it actually breaks. The derive approach is more of an opposite. And after this discussion I'd rather lean on the side of being explicit.